File: apps/cli/Services/FileScanner.cs
using System.Diagnostics;
using DepVault.Cli.Utils;

namespace DepVault.Cli.Services;

public enum FileCategory
{
    Dependency,
    Environment,
    SecretFile
}

public record DiscoveredFile(string FullPath, string RelativePath, string FileName, FileCategory Category);

public interface IFileScanner
{
    List<DiscoveredFile> FindDependencyFiles(string rootPath);
    List<DiscoveredFile> FindEnvFiles(string rootPath);
    List<DiscoveredFile> FindSecretFiles(string rootPath);
    List<DiscoveredFile> FindAllPushableFiles(string rootPath);
    FileCategory ClassifyFile(string fileName);
}

public sealed class FileScanner : IFileScanner
{
    private static readonly IReadOnlySet<string> excludedDirs = ExcludedDirectories.Names;

    private static readonly HashSet<string> envFileNames = new(StringComparer.OrdinalIgnoreCase)
    {
        ".env", ".env.local", ".env.development", ".env.staging", ".env.production",
        ".env.test"
    };

    private static readonly HashSet<string> secretFileExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".pem", ".key", ".p12", ".pfx", ".jks", ".keystore"
    };

    private static readonly HashSet<string> secretFileNames = new(StringComparer.OrdinalIgnoreCase)
    {
        "firebase-config.json", "service-account.json", "service-account-key.json",
        "credentials.json", "gcp-key.json", "google-services.json",
        "GoogleService-Info.plist"
    };

    private static readonly string[] secretFileNamePatterns = ["firebase-adminsdk"];

    public List<DiscoveredFile> FindDependencyFiles(string rootPath)
    {
        return ScanFiles(rootPath, FileCategory.Dependency, IsDependencyFile);
    }

    public List<DiscoveredFile> FindEnvFiles(string rootPath)
    {
        var files = ScanFiles(rootPath, FileCategory.Environment, IsEnvFile);
        return FilterGitTrackedConfigFiles(rootPath, files);
    }

    public List<DiscoveredFile> FindSecretFiles(string rootPath)
    {
        return ScanFiles(rootPath, FileCategory.SecretFile, IsSecretFile);
    }

    public List<DiscoveredFile> FindAllPushableFiles(string rootPath)
    {
        return FindEnvFiles(rootPath)
            .Concat(FindSecretFiles(rootPath))
            .OrderBy(f => f.RelativePath)
            .ToList();
    }

    public FileCategory ClassifyFile(string fileName)
    {
        if (IsEnvFile(fileName))
        {
            return FileCategory.Environment;
        }

        if (IsSecretFile(fileName))
        {
            return FileCategory.SecretFile;
        }

        return FileCategory.SecretFile;
    }

    private static List<DiscoveredFile> ScanFiles(string rootPath, FileCategory category, Func<string, bool> predicate)
    {
        var results = new List<DiscoveredFile>();
        var fullRoot = Path.GetFullPath(rootPath);

        try
        {
            var files = Directory.EnumerateFiles(fullRoot, "*", new EnumerationOptions
            {
                RecurseSubdirectories = true,
                IgnoreInaccessible = true,
                MaxRecursionDepth = 10
            });

            foreach (var filePath in files)
            {
                if (IsInExcludedDirectory(filePath, fullRoot))
                {
                    continue;
                }

                var fileName = Path.GetFileName(filePath);
                if (predicate(fileName))
                {
                    var relativePath = Path.GetRelativePath(fullRoot, filePath).Replace('\\', '/');
                    results.Add(new DiscoveredFile(filePath, relativePath, fileName, category));
                }
            }
        }
        catch (UnauthorizedAccessException)
        {
        }
        catch (DirectoryNotFoundException)
        {
        }

        return results.OrderBy(f => f.RelativePath).ToList();
    }

    /// <summary>
    /// Filters out files in excluded directories like node_modules, vendor, .git, etc.
    /// This is important to avoid suggesting irrelevant files for push/pull and to prevent accidental large uploads.
    /// </summary>
    private static bool IsInExcludedDirectory(string filePath, string rootPath)
    {
        var relativePath = Path.GetRelativePath(rootPath, filePath);
        var parts = relativePath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);

        for (var i = 0; i < parts.Length - 1; i++)
        {
            if (excludedDirs.Contains(parts[i]))
            {
                return true;
            }
        }

        return false;
    }

    private static bool IsDependencyFile(string fileName)
    {
        return EcosystemResolver.IsDependencyFile(fileName);
    }

    private static bool IsEnvFile(string fileName)
    {
        if (envFileNames.Contains(fileName))
        {
            return true;
        }

        if (fileName.StartsWith(".env.", StringComparison.OrdinalIgnoreCase) &&
            !fileName.EndsWith(".bak", StringComparison.OrdinalIgnoreCase) &&
            !fileName.EndsWith(".example", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (fileName.Equals("appsettings.json", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (fileName.StartsWith("appsettings.", StringComparison.OrdinalIgnoreCase) &&
            fileName.EndsWith(".json", StringComparison.OrdinalIgnoreCase) &&
            fileName.Count(c => c == '.') == 2)
        {
            return true;
        }

        return fileName.Equals("secrets.yaml", StringComparison.OrdinalIgnoreCase)
               || fileName.Equals("secrets.yml", StringComparison.OrdinalIgnoreCase)
               || fileName.Equals("config.toml", StringComparison.OrdinalIgnoreCase);
    }

    private static bool IsSecretFile(string fileName)
    {
        var ext = Path.GetExtension(fileName);
        if (secretFileExtensions.Contains(ext))
        {
            return true;
        }

        if (secretFileNames.Contains(fileName))
        {
            return true;
        }

        return secretFileNamePatterns.Any(p => fileName.Contains(p, StringComparison.OrdinalIgnoreCase));
    }

    /// <summary>
    /// Filters out config files (appsettings.json, secrets.yaml, config.toml) that are tracked
    /// by git. Tracked config files typically contain placeholders, not real secrets.
    /// Only gitignored/untracked config files are suggested for push/pull.
    /// Plain .env files are always included regardless of git status.
    /// </summary>
    private static List<DiscoveredFile> FilterGitTrackedConfigFiles(string rootPath, List<DiscoveredFile> files)
    {
        var configFiles = files.Where(f => IsConfigStyleFile(f.FileName)).ToList();
        if (configFiles.Count == 0)
        {
            return files;
        }

        var gitIgnored = GetGitIgnoredPaths(rootPath, configFiles);

        return files
            .Where(f => !IsConfigStyleFile(f.FileName) || gitIgnored.Contains(f.RelativePath))
            .ToList();
    }

    private static bool IsConfigStyleFile(string fileName)
    {
        return fileName.StartsWith("appsettings", StringComparison.OrdinalIgnoreCase)
               || fileName.Equals("secrets.yaml", StringComparison.OrdinalIgnoreCase)
               || fileName.Equals("secrets.yml", StringComparison.OrdinalIgnoreCase)
               || fileName.Equals("config.toml", StringComparison.OrdinalIgnoreCase);
    }

    private static HashSet<string> GetGitIgnoredPaths(string rootPath, List<DiscoveredFile> files)
    {
        var ignored = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        try
        {
            var paths = string.Join('\n', files.Select(f => f.RelativePath));
            var psi = new ProcessStartInfo("git", "check-ignore --stdin")
            {
                WorkingDirectory = rootPath,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using var process = Process.Start(psi);
            if (process is null)
            {
                return ignored;
            }

            process.StandardInput.Write(paths);
            process.StandardInput.Close();

            var output = process.StandardOutput.ReadToEnd();
            process.WaitForExit(5000);

            foreach (var line in output.Split('\n', StringSplitOptions.RemoveEmptyEntries))
            {
                ignored.Add(line.Trim().Replace('\\', '/'));
            }
        }
        catch
        {
            // Not a git repo or git not available — include all files
            foreach (var f in files)
            {
                ignored.Add(f.RelativePath);
            }
        }

        return ignored;
    }
}

File: apps/cli-ts/src/services/file-scanner.ts
import { readdirSync, statSync } from "node:fs";
import { basename, extname, join, relative } from "node:path";
import { DEPENDENCY_FILE_MAP } from "@depvault/shared";

const MAX_DEPTH = 10;

const EXCLUDED_DIRS = new Set([
  ".git",
  ".hg",
  ".svn",
  "node_modules",
  ".yarn",
  ".pnpm-store",
  "bower_components",
  "__pycache__",
  ".venv",
  "venv",
  ".tox",
  ".mypy_cache",
  ".pytest_cache",
  ".ruff_cache",
  "bin",
  "obj",
  ".nuget",
  "packages",
  "TestResults",
  "publish",
  "artifacts",
  "target",
  "vendor",
  ".gradle",
  ".mvn",
  ".bundle",
  "dist",
  "build",
  "out",
  ".output",
  ".next",
  ".angular",
  ".turbo",
  ".nx",
  ".parcel-cache",
  ".cache",
  ".vite",
  ".svelte-kit",
  ".vs",
  ".idea",
  ".vscode",
  "coverage",
  "generated",
  "logs",
  "temp",
  "tmp",
]);

const ENV_FILE_NAMES = new Set([
  ".env",
  ".env.local",
  ".env.development",
  ".env.staging",
  ".env.production",
  ".env.test",
]);

const SECRET_FILE_EXTENSIONS = new Set([".pem", ".key", ".p12", ".pfx", ".jks", ".keystore"]);

const SECRET_FILE_NAMES = new Set([
  "firebase-config.json",
  "service-account.json",
  "service-account-key.json",
  "credentials.json",
  "gcp-key.json",
  "google-services.json",
  "GoogleService-Info.plist",
]);

function walk(dir: string, depth: number, results: string[]): void {
  if (depth > MAX_DEPTH) return;

  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }

  for (const entry of entries) {
    if (EXCLUDED_DIRS.has(entry.toLowerCase())) continue;

    const fullPath = join(dir, entry);
    let stat;
    try {
      stat = statSync(fullPath);
    } catch {
      continue;
    }

    if (stat.isDirectory()) {
      walk(fullPath, depth + 1, results);
    } else if (stat.isFile()) {
      results.push(fullPath);
    }
  }
}

function getAllFiles(rootDir: string): string[] {
  const results: string[] = [];
  walk(rootDir, 0, results);
  return results;
}

export function findDependencyFiles(rootDir: string): string[] {
  return getAllFiles(rootDir).filter((f) => {
    const name = basename(f);
    return name in (DEPENDENCY_FILE_MAP as Record<string, unknown>);
  });
}

export function findEnvFiles(rootDir: string): string[] {
  return getAllFiles(rootDir).filter((f) => {
    const name = basename(f);
    return ENV_FILE_NAMES.has(name) || name.startsWith(".env.");
  });
}

export function findSecretFiles(rootDir: string): string[] {
  return getAllFiles(rootDir).filter((f) => {
    const name = basename(f);
    const ext = extname(f).toLowerCase();
    return (
      SECRET_FILE_EXTENSIONS.has(ext) ||
      SECRET_FILE_NAMES.has(name) ||
      name.includes("firebase-adminsdk")
    );
  });
}

export type FileCategory = "environment" | "secret";

export interface DiscoveredFile {
  fullPath: string;
  relativePath: string;
  fileName: string;
  category: FileCategory;
}

/** Find all pushable files (env + secret) and return them with metadata. */
export function findPushableFiles(rootDir: string): DiscoveredFile[] {
  const envFiles = findEnvFiles(rootDir).map((f) => toDiscovered(rootDir, f, "environment"));
  const secretFiles = findSecretFiles(rootDir).map((f) => toDiscovered(rootDir, f, "secret"));
  return [...envFiles, ...secretFiles].sort((a, b) => a.relativePath.localeCompare(b.relativePath));
}

function toDiscovered(rootDir: string, fullPath: string, category: FileCategory): DiscoveredFile {
  return {
    fullPath,
    relativePath: relative(rootDir, fullPath).replace(/\\/g, "/"),
    fileName: basename(fullPath),
    category,
  };
}

/** Detect environment type from a filename (e.g. ".env.production" → "PRODUCTION"). */
export function detectEnvironmentType(fileName: string): string | null {
  const segments = fileName.toLowerCase().split(/[.\-_]/);
  for (const seg of segments) {
    if (seg === "production" || seg === "prod") return "PRODUCTION";
    if (seg === "staging" || seg === "stage") return "STAGING";
    if (seg === "development" || seg === "dev") return "DEVELOPMENT";
  }
  return null;
}

/** Detect config format from filename extension. */
export function detectFormat(fileName: string): string {
  const lower = fileName.toLowerCase();
  if (lower.endsWith(".json")) return "appsettings.json";
  if (lower.endsWith(".yaml") || lower.endsWith(".yml")) return "secrets.yaml";
  if (lower.endsWith(".toml")) return "config.toml";
  return "env";
}
